pyERGM/metrics.py

```diff
@@ -279,16 +340,68 @@ class BaseDegreeVector(Metric):
     which indices the calculation ignores.
     """
 
-    def __init__(self, requires_graph: bool, is_directed: bool, indices_from_user=None):
-        super().__init__(requires_graph=requires_graph)
+    def __init__(self, is_directed: bool, indices_from_user=None):
+        super().__init__(requires_graph=False)
 
         self._indices_from_user = indices_from_user.copy() if indices_from_user is not None else None
 
         self._is_directed = is_directed
 
-    def _get_total_feature_count(self): 
+    def _get_total_feature_count(self):
         return self._n_nodes
 
+    def calculate_bootstrapped_features(self, first_halves_to_use: np.ndarray,
+                                        second_halves_to_use: np.ndarray,
+                                        first_halves_indices: np.ndarray[int], second_halves_indices: np.ndarray[int]):
+        """
+        Calculates the bootstrapped degree, by counting the connections of each node in the sampled subnetworks, and
+        normalizing by network size (i.e., calculating the fraction of existing edges out of all possible ones for each
+        node in sampled subnetworks, and multiplying by the number of possible edges of a single node in the full
+        observed network).
+        Each node appears in one of the sub-samples, so both are used, and the indices are used to order the calculated
+        values in the entire features vector.
+        Parameters
+        ----------
+        first_halves_to_use
+            Multiple samples of subnetworks of an observed network, representing the connectivity between half of the
+            nodes in the large network.
+        second_halves_to_use
+            The subnetworks formed by the complementary set of nodes of the large network for each sample.
+        first_halves_indices
+            The indices of the nodes in the first half of the large network for each sample, according to the ordering
+            of the nodes in the large network.
+        second_halves_indices
+            The indices of the nodes in the second half of the large network for each sample, according to the ordering
+
+        Returns
+        -------
+        Properly normalized statistics of subnetworks of an observed network.
+        """
+        num_nodes_in_observed = first_halves_indices.shape[0] + second_halves_indices.shape[0]
+        num_nodes_in_first_half = first_halves_indices.shape[0]
+        num_nodes_in_second_half = second_halves_indices.shape[0]
+
+        # NOTE! we want the estimated covariance matrix from the bootstrap to match the dimension of samples from the
+        # model during optimization, so we must first calculate the degrees for all the nodes (thus the initialization
+        # of a new instance, which makes sure that no indices are ignored), and then ignore the indices that will be
+        # removed by the metric.
+        fresh_instance = self.__class__()  # No child class has a `is_directed` input to the constructor.
```

Comment:
nice idea. but I don't understand the scenario for `UndirectedDegree`. 
It needs  `is_directed=False`, so what's going to happen in `fresh_instance = self.__class__()` ?

Reply:
This is the default of the class, so this is what will be passed. Actually, I don't know why we pass it to the constructor (also for the In/Out degrees), we should just set it inside as it can't really have another value.
There was another annoying issue with `UndirectedDegree` here, let's discuss that